### PR TITLE
Support backprop in `default.mixed` device for PyTorch

### PR DESCRIPTION
**Pull request by eddddddy**
**Context:**
The third PR in a series to support backpropagation in the `default.mixed` device.

**Description of the Change:**
Now backpropagation is supported in the `default.mixed` device for the PyTorch interface.

**Benefits:**
Better integration with auto-differentiation frameworks.

**Possible Drawbacks:**

**Related GitHub Issues:**
#1528

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -407,10 +407,10 @@
   tensor([0.69301172, 0.67552491, 0.65128847], requires_grad=True)
   ```
 
-* The `default.mixed` device now supports backpropagation with the Autograd and TensorFlow
-  interfaces.
+* The `default.mixed` device now supports backpropagation with the Autograd, TensorFlow, and PyTorch interfaces.
```
Comment (antalszava) [inline doc/releases/changelog-dev.md:410]:
```suggestion
* The `default.mixed` device now supports backpropagation with the Autograd, TensorFlow, and PyTorch (CPU) interfaces.
```

File: pennylane/math/single_dispatch.py [hunk 2]
```diff
@@ -372,11 +372,30 @@ def _to_numpy_torch(x):
...
 
 ar.register_function("torch", "to_numpy", _to_numpy_torch)
-ar.register_function(
-    "torch",
-    "asarray",
-    lambda x, device=None: _i("torch").as_tensor(x, device=device),
-)
+
+
+def _asarray_torch(x, dtype=None, device=None):
```
Comment (antalszava) [inline pennylane/math/single_dispatch.py:377]: The signature of this function seems to not match completely the signature of `torch.asarray`. How come?

**Review by antalszava — APPROVED** [review]
Looks good @eddddddy! :tada: :partying_face: 

One main feedback is the fact that users might not able to specify Torch devices in a versatile manner, as they would with `default.qubit.torch`. Having said that, it may be worth merging this as is because it should work for CPU and then another PR could address adding GPU support. This should just be documented and explained well (changelog and/or device docs).

File: tests/devices/test_default_mixed_torch.py [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+from pennylane import numpy as np
+from pennylane.devices.default_mixed import DefaultMixed
+from pennylane import DeviceError
+
+pytestmark = pytest.mark.torch
+
+torch = pytest.importorskip("torch")
+
+
+class TestQNodeIntegration:
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply tests/devices/test_default_mixed_torch.py:30]: We already set `pytestmark = pytest.mark.torch` at the start. I believe this marks everything in the file.

File: pennylane/math/single_dispatch.py (hunk 2 above)
Comment (eddddddy) [reply pennylane/math/single_dispatch.py:377]: Since it's a `qml.math` dispatch function, I think it should more closely match the signatures of `np.asarray`, `tf.convert_to_tensor`, etc.

File: pennylane/math/single_dispatch.py [hunk 4]
```diff
@@ -372,11 +372,30 @@ def _to_numpy_torch(x):
...
+        np.int32: torch.int32,
+        np.int64: torch.int64,
+        np.float16: torch.float16,
+        np.float32: torch.float32,
+        np.float64: torch.float64,
+        np.complex64: torch.complex64,
+        np.complex128: torch.complex128,
+    }
+
+    if dtype in dtype_map:
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/math/single_dispatch.py:392]: Depends on the behaviour we want. Both Autograd and TensorFlow allows NumPy dtypes in tensors (which gets converted automatically into the interface-specific dtypes), but PyTorch doesn't. I tried to be as consistent as possible across the interfaces and allow passing in NumPy dtypes for all.

File: tests/devices/test_default_mixed_torch.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (antalszava) [reply tests/devices/test_default_mixed_torch.py:30]: Oh definitely, missed that one! My bad. :+1:

File: pennylane/math/single_dispatch.py (hunk 2 above)
Comment (eddddddy) [reply pennylane/math/single_dispatch.py]: Allowed extra `kwargs` in this function which are passed to `torch.asarray`.

File: pennylane/math/single_dispatch.py [hunk 5]
```diff
@@ -533,6 +562,22 @@ def _ndim_torch(tensor):
 ar.register_function("torch", "eigvalsh", lambda x: _i("torch").linalg.eigvalsh(x))
 ar.register_function("torch", "entr", lambda x: _i("torch").sum(_i("torch").special.entr(x)))
 
+
+def _sum_torch(tensor, axis=None, keepdims=False):
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/math/single_dispatch.py]: Only reason is I didn't know that `sum` took a `dtype` argument 😄 . Added this argument to the function.

File: pennylane/math/single_dispatch.py [hunk 6]
```diff
@@ -492,6 +511,15 @@ def _block_diag_torch(tensors):
 ar.register_function("torch", "block_diag", _block_diag_torch)
 
 
+def _scatter_torch(indices, array, new_dimensions):
+    import torch
+
+    tensor = array
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/math/single_dispatch.py]: No reason, changed in most recent commit.

**Comment by antalszava** [comment]
[sc-21103]
